lib/src/css_mirroring/rulesets_processor.dart

```diff
@@ -0,0 +1,82 @@
+part of scissors.src.css_mirroring.bidi_css_generator;
+
+abstract class NodeProcessor {
+  void process(EditConfiguration e, TopLevelsPair t, int i);
+}
+
+class DeclarationProcessor extends NodeProcessor {
+  final _PendingRemovals removableRuleSets;
+  final int parentEnd;
+
+  DeclarationProcessor(this.removableRuleSets, this.parentEnd);
+
+  @override
+  void process(EditConfiguration _editConfig, TopLevelsPair _topLevelsPair,
+      int _iTopLevel) {
+    final removals = new _PendingRemovals(_editConfig.transaction);
+    final List<TreeNode> usedTopLevels =
+        _getUsedTopLevels(_editConfig.mode, _topLevelsPair);
+
+    _storeRemovableDeclarations(
+        _editConfig, _topLevelsPair, _iTopLevel, removals);
+
+    if (_isRuleRemovable(
+        removals, _topLevelsPair._originalTopLevels[_iTopLevel])) {
+      removableRuleSets.addRemoval(_getNodeStart(usedTopLevels[_iTopLevel]),
+          _getRuleSetEnd(usedTopLevels, _iTopLevel, parentEnd));
+    } else {
+      removals.commitRemovals();
+
+      /// Add direction attribute to RuleId for direction-specific RuleSet.
+      if (_editConfig.mode != _RetentionMode.keepBidiNeutral) {
+        _prependDirectionToRuleSet(_editConfig, usedTopLevels[_iTopLevel]);
+      }
+    }
+  }
+
+  /// Stores start and end locations of removable declarations in a ruleset
+  /// based upon the retention mode.
+  void _storeRemovableDeclarations(
```

Comment:
nothing actually edits or remove until we commit, that's a general principle everywhere (on pending removals, on a text transaction, etc). My point is "store removable declarations" seems *weird* (where are they stored? removable or to be removed? etc). Remove declarations says it removes declarations, which is what it does (the pending removals arg makes it clear the removal is gonna be pending)